src/Castle.Core.AsyncInterceptor/InterceptorAdapter.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Copyright (c) 2016 James Skimming. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See LICENSE in the project root for license information.
+
+namespace Castle.DynamicProxy
+{
+    using System.Reflection;
+    using System.Runtime.CompilerServices;
+    using System.Runtime.ExceptionServices;
+    using System.Threading.Tasks;
+
+    using Castle.DynamicProxy.Invocations;
+
+    /// <inheritdoc cref="IInterceptor"/>
+    internal class InterceptorAdapter : IInterceptor
+    {
+        private static readonly MethodInfo InterceptAsyncFunctionMethodInfo = typeof(InterceptorAdapter).GetMethod(
+            nameof(ExecuteForAsyncFunction),
+            BindingFlags.Instance | BindingFlags.NonPublic);
+
+        private static readonly MethodInfo InterceptFunctionMethodInfo = typeof(InterceptorAdapter).GetMethod(
+            nameof(ExecuteForFunction),
+            BindingFlags.Instance | BindingFlags.NonPublic);
+
+        private readonly IAsyncInterceptor _interceptor;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="InterceptorAdapter"/> class.
+        /// </summary>
+        /// <param name="interceptor">
+        /// The interceptors.
+        /// </param>
+        public InterceptorAdapter(IAsyncInterceptor interceptor)
+        {
+            _interceptor = interceptor;
+        }
+
+        /// <summary>
+        /// Intercepts a method <paramref name="invocation"/>.
+        /// </summary>
+        /// <param name="invocation">The method invocation.</param>
+        public void Intercept(IInvocation invocation)
+        {
+            try
```

Comment:
Personally previously (with GetMethodType helper and MethodType enum) it was easier to read this method

Reply:
Yes,  that makes sense. Good call